pkgs/tools/typesetting/asciidoctor/default.nix

```
@@ -1,59 +1,29 @@
-{ stdenv, lib, bundlerApp, ruby, bundix, mkShell
-  # Dependencies of the 'mathematical' package
-, cmake, bison, flex, glib, pkgconfig, cairo
-, pango, gdk_pixbuf, libxml2, python3, patchelf
-}:
+{ lib, bundlerApp, mkShell, bundix, defaultGemConfig }:
 
-bundlerApp rec {
-  inherit ruby;
-  pname = "asciidoctor";
-  gemdir = ./.;
+let app = bundlerApp {
+    pname = "asciidoctor";
+    gemdir = ./.;
 
-  exes = [
-    "asciidoctor"
-    "asciidoctor-pdf"
-    "asciidoctor-safe"
-  ];
-
-  gemConfig = {
-    mathematical = attrs: {
-      buildInputs = [
-        cmake
-        bison
-        flex
-        glib
-        pkgconfig
-        cairo
-        pango
-        gdk_pixbuf
-        libxml2
-        python3
-      ];
-
-      # The ruby build script takes care of this
-      dontUseCmakeConfigure = true;
+    exes = [
+      "asciidoctor"
+      "asciidoctor-pdf"
+      "asciidoctor-safe"
+      "asciidoctor-epub3"
+    ];
 
-      # For some reason 'mathematical.so' is missing cairo and glib in its RPATH, add them explicitly here
-      postFixup = lib.optionalString stdenv.isLinux ''
-        soPath="$out/${ruby.gemPath}/gems/mathematical-${attrs.version}/lib/mathematical/mathematical.so"
-        ${patchelf}/bin/patchelf \
-          --set-rpath "${lib.makeLibraryPath [ glib cairo ]}:$(${patchelf}/bin/patchelf --print-rpath "$soPath")" \
-          "$soPath"
-      '';
+    meta = with lib; {
+      description = "A faster Asciidoc processor written in Ruby";
+      homepage = https://asciidoctor.org/;
+      license = licenses.mit;
+      maintainers = with maintainers; [ gpyh ];
+      platforms = platforms.unix;
     };
   };
 
-  passthru.updateShell = mkShell {
-    buildInputs = (gemConfig.mathematical {}).buildInputs ++ [
-      bundix
-    ];
-  };
-
-  meta = with lib; {
-    description = "A faster Asciidoc processor written in Ruby";
-    homepage = https://asciidoctor.org/;
-    license = licenses.mit;
-    maintainers = with maintainers; [ gpyh ];
-    platforms = platforms.unix;
+  shell = mkShell { 
+    buildInputs = 
+      app.gems.mathematical.buildInputs ++ 
+      app.gems.nokogiri.buildInputs ++ 
+      [ bundix ]; 
   };
-}
+in app.overrideAttrs (attrs: { passthru = attrs.passthru // { updateShell = shell; }; })
```

```suggestion
in app
```
attach the shell directly on the app.passthru to avoid the overhead

---

This doesn't work as `app.gems` isn't defined then.

---

I pushed f726160556e74c8203f3f8dff0aa0dec07c0218c which demonstrates how to implement it. The key is that nix is a lazily-evaluated language.